File: fortran/heat1d.f90
! 1D Heat equation: u_t = alpha*u_xx
! Domain: [0,1], BC: u=0, IC: sin(pi*x)
! Exact: exp(-alpha*pi^2*t)*sin(pi*x)
! Method: explicit FTCS
! Output: heat1d_result.csv

program heat1d
    implicit none

    real(8), parameter :: alpha = 1.0d0
    integer, parameter :: N     = 50
    real(8), parameter :: T_END = 0.1d0
    real(8), parameter :: R     = 0.4d0   ! CFL: alpha*dt/dx^2

    real(8) :: dx, dt, t_final, l2, e
    integer :: M, steps, n, i
    real(8), allocatable :: x(:), u(:), u_new(:), u_ex(:)

    real(8), parameter :: PI = acos(-1.0d0)

    M  = N + 2
    dx = 1.0d0 / (N + 1)
    dt = R * dx * dx / alpha
    steps = int(T_END / dt) + 1

    allocate(x(M), u(M), u_new(M), u_ex(M))

    do i = 1, M
        x(i) = (i-1) * dx
        u(i) = sin(PI * x(i))
    end do
    u(1) = 0.0d0; u(M) = 0.0d0

    ! time march
    do n = 1, steps
        do i = 2, M-1
            u_new(i) = u(i) + R*(u(i+1) - 2.0d0*u(i) + u(i-1))
        end do
        u_new(1) = 0.0d0; u_new(M) = 0.0d0
        u = u_new
    end do

    ! exact and L2 error
    t_final = (steps - 1) * dt
    l2 = 0.0d0
    do i = 1, M
        u_ex(i) = exp(-alpha * PI**2 * t_final) * sin(PI * x(i))
        e = u(i) - u_ex(i)
        l2 = l2 + e*e * dx
    end do
    l2 = sqrt(l2)
    write(*,'(A,I4,A,F6.3,A,ES10.3)') &
        '1D Heat  N=', N, '  T=', T_END, '  L2 error = ', l2

    ! save CSV
    open(unit=10, file='heat1d_result.csv', status='replace')
    write(10,'(A)') 'x,u_numerical,u_exact,error'
    do i = 1, M
        write(10,'(4(ES14.6,","))') x(i), u(i), u_ex(i), u(i)-u_ex(i)
    end do
    close(10)
    write(*,*) 'Saved heat1d_result.csv'

    deallocate(x, u, u_new, u_ex)
end program heat1d

File: cpp/heat1d.cpp
/*
 * 1D Heat equation: u_t = alpha * u_xx
 * Domain: [0,1], BC: u=0, IC: sin(pi*x)
 * Exact: exp(-alpha*pi^2*t)*sin(pi*x)
 * Method: explicit FTCS
 * Output: heat1d_result.csv (columns: x, u_numerical, u_exact, error)
 */

#include <cmath>
#include <cstdio>
#include <vector>
#include <string>
#include <fstream>

static const double PI    = std::acos(-1.0);
static const double ALPHA = 1.0;
static const int    N     = 50;
static const double T_END = 0.1;
static const double R     = 0.4;     // CFL: alpha*dt/dx^2

void save_csv(const std::string& fname,
              const std::vector<double>& x,
              const std::vector<double>& u_num,
              const std::vector<double>& u_ex)
{
    std::ofstream f(fname);
    f << "x,u_numerical,u_exact,error\n";
    for (int i = 0; i <= N+1; ++i)
        f << x[i] << "," << u_num[i] << "," << u_ex[i]
          << "," << (u_num[i]-u_ex[i]) << "\n";
}

int main()
{
    int M = N + 2;
    std::vector<double> x(M), u(M), u_new(M);

    double dx = 1.0 / (N + 1);
    for (int i = 0; i < M; ++i) x[i] = i * dx;

    double dt    = R * dx * dx / ALPHA;
    int    steps = static_cast<int>(T_END / dt) + 1;

    // initial condition
    for (int i = 0; i < M; ++i)
        u[i] = std::sin(PI * x[i]);
    u[0] = u[M-1] = 0.0;

    // time march
    for (int n = 0; n < steps; ++n) {
        for (int i = 1; i <= N; ++i)
            u_new[i] = u[i] + R * (u[i+1] - 2.0*u[i] + u[i-1]);
        u_new[0] = u_new[M-1] = 0.0;
        std::swap(u, u_new);
    }

    // exact solution and L2 error
    double t_final = (steps - 1) * dt;
    std::vector<double> u_ex(M);
    double l2 = 0.0;
    for (int i = 0; i < M; ++i) {
        u_ex[i] = std::exp(-ALPHA * PI * PI * t_final) * std::sin(PI * x[i]);
        double e = u[i] - u_ex[i];
        l2 += e * e * dx;
    }
    l2 = std::sqrt(l2);
    std::printf("1D Heat  N=%d  T=%.2f  L2 error = %.3e\n", N, T_END, l2);

    save_csv("heat1d_result.csv", x, u, u_ex);
    std::printf("Saved heat1d_result.csv\n");
    return 0;
}
